Parallelize the `provision` step

happz opened the pull request:
This change requires refactoring of current queue internals, to better support the third kind of parallelization & queueing. A custom queue task is needed, to run provisioning in parallel, and the step takes care of ordering of provision phases and actions.

Implements https://github.com/teemtee/tmt/issues/2244

Pull Request Checklist

* [x] implement the feature
* [x] include a release note

lukaszachy commented:
Hm, I've tried with virtual and it doesn't work well.

```
/test:
    test: echo
    
/plan:
    execute:
        how: tmt
    discover:
        how: fmf
        
    provision:
        - how: virtual
        - how: virtual
        - how: virtual
```

```
$ tmt run --until report
/var/tmp/tmt/run-117

/p/plan
    discover
        how: fmf
        directory: /tmp/r_173145_H6I
        summary: 1 test selected
    provision
        queued provision task #1: default-0, default-1 and default-2
        
        provision task #1: default-0, default-1 and default-2
[default-0]         started
[default-0]         how: virtual
[default-1]         started
[default-0]         memory: 2048 megabyte
[default-0]         disk: 40 gigabyte
[default-1]         how: virtual
[default-2]         started
[default-1]         memory: 2048 megabyte
[default-1]         disk: 40 gigabyte
[default-2]         how: virtual
[default-2]         memory: 2048 megabyte
[default-2]         disk: 40 gigabyte
[default-0]         progress: booting...
[default-1]         progress: booting...
[default-2]         progress: booting...
libvirt: QEMU Driver error : internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:32.338285Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22'
Instance startup failed, retrying in 5 seconds...
libvirt: QEMU Driver error : internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22'
[default-0]         finished
[default-0]         fail: Failed to boot testcloud instance (internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22').
[default-2]         finished
[default-2]         multihost name: default-2
[default-2]         arch: x86_64
[default-2]         distro: Fedora Linux 39 (Cloud Edition)
[default-1]         finished
[default-1]         multihost name: default-1
[default-1]         arch: x86_64
[default-1]         distro: Fedora Linux 39 (Cloud Edition)

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    provision step failed

    The exception was caused by 1 earlier exceptions

    Cause number 1:

        Failed to boot testcloud instance (internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22').
```

There should be 3 machines running but virsh shows that one of them is paused.

lukaszachy commented on tmt/steps/prepare/__init__.py:
```diff
@@ -247,7 +254,14 @@ def go(self) -> None:
+                PushTask(
+                    logger=self._logger,
+                    result=None,
+                    guest=None,
```
`guest` and `guests` ... Could they be renamed? Highly confusing to have them both in single signature

happz replied:
Right, I forgot it's still ugly. I'll improve this, can't use the default `__init__` and default values.

psss replied:
Agreed, would be better to make the difference more clear.

happz replied:
Hopefully improved in 69824309

psss replied:
Yeap, thanks!

happz commented:
> Hm, I've tried with virtual and it doesn't work well.
>
> ...
> libvirt: QEMU Driver error : internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:32.338285Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22'
> Instance startup failed, retrying in 5 seconds...
> libvirt: QEMU Driver error : internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22'
> [default-0]         finished
> [default-0]         fail: Failed to boot testcloud instance (internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22').
> [default-2]         finished
> [default-2]         multihost name: default-2
> [default-2]         arch: x86_64
> [default-2]         distro: Fedora Linux 39 (Cloud Edition)
> [default-1]         finished
> [default-1]         multihost name: default-1
> [default-1]         arch: x86_64
> [default-1]         distro: Fedora Linux 39 (Cloud Edition)
> 
> plan failed
> 
> The exception was caused by 1 earlier exceptions
> 
> Cause number 1:
> 
>     provision step failed
> 
>     The exception was caused by 1 earlier exceptions
> 
>     Cause number 1:
> 
>         Failed to boot testcloud instance (internal error: QEMU unexpectedly closed the monitor (vm='tmt-117-idTZSMZI'): 2023-11-21T16:35:37.575292Z qemu-system-x86_64: -netdev user,id=testcloud_net.10042,hostfwd=tcp::10042-:22: Could not set up host forwarding rule 'tcp::10042-:22').
> ```
> 
> There should be 3 machines running but virsh shows that one of them is paused.

Worked like a charm on my laptop :/ Can you check whether there's some libvirt/qemu/testcloud logging related to this? I'm not versed in Qemu things, but it looks like some kind of conflict. @frantisekz any idea what might be we hitting?

lukaszachy commented:
@happz I'd say https://pagure.io/testcloud/blob/master/f/testcloud/util.py#_60 is not ready for parallel requests.

frantisekz commented:
I'll take a look early December (currently on PTO without laptop).

happz commented:
> @happz I'd say https://pagure.io/testcloud/blob/master/f/testcloud/util.py#_60 is not ready for parallel requests.

Right, that might be the case, easily.

Can you try once more with the following patch? It should prevent concurrent calls into testcloud library:

```diff
diff --git a/tmt/steps/provision/testcloud.py b/tmt/steps/provision/testcloud.py
index 02b9453b..79776ec4 100644
--- a/tmt/steps/provision/testcloud.py
+++ b/tmt/steps/provision/testcloud.py
@@ -4,6 +4,7 @@ import datetime
 import os
 import platform
 import re
+import threading
 import time
 import types
 from typing import TYPE_CHECKING, Any, Optional, Union, cast
@@ -336,6 +337,11 @@ class GuestTestcloud(tmt.GuestSsh):
     _domain: Optional[  # type: ignore[name-defined]
         'testcloud.domain_configuration.DomainConfiguration'] = None
 
+    #: The lock protects calls into the testcloud library. We suspect it might
+    #: be unprepared for multi-threaded use. After the dust settles, we may
+    #: remove the lock.
+    _testcloud_lock = threading.Lock()
+
     @property
     def is_ready(self) -> bool:
         if self._instance is None:
@@ -343,14 +349,17 @@ class GuestTestcloud(tmt.GuestSsh):
 
         assert testcloud is not None
         assert libvirt is not None
-        try:
-            state = testcloud.instance._find_domain(self._instance.name, self._instance.connection)
-            # Note the type of variable 'state' is 'Any'. Hence, we don't use:
-            #     return state == 'running'
-            # to avoid error from type checking.
-            return bool(state == "running")
-        except libvirt.libvirtError:
-            return False
+
+        with GuestTestcloud._testcloud_lock:
+            try:
+                state = testcloud.instance._find_domain(
+                    self._instance.name, self._instance.connection)
+                # Note the type of variable 'state' is 'Any'. Hence, we don't use:
+                #     return state == 'running'
+                # to avoid error from type checking.
+                return bool(state == "running")
+            except libvirt.libvirtError:
+                return False
 
     def _get_url(self, url: str, message: str) -> requests.Response:
         """ Get url, retry when fails, return response """
@@ -407,9 +416,11 @@ class GuestTestcloud(tmt.GuestSsh):
         self._image = testcloud.image.Image(self.image_url)
         if self.instance_name is None:
             raise ProvisionError(f"The instance name '{self.instance_name}' is invalid.")
-        self._instance = testcloud.instance.Instance(
-            self.instance_name, image=self._image,
-            connection=f"qemu:///{self.connection}", desired_arch=self.arch)
+
+        with GuestTestcloud._testcloud_lock:
+            self._instance = testcloud.instance.Instance(
+                self.instance_name, image=self._image,
+                connection=f"qemu:///{self.connection}", desired_arch=self.arch)
 
     def prepare_ssh_key(self, key_type: Optional[str] = None) -> None:
         """ Prepare ssh key for authentication """
@@ -700,11 +711,12 @@ class GuestTestcloud(tmt.GuestSsh):
             seed_disk = RawStorageDevice(self._domain.seed_path)
             self._domain.storage_devices.append(seed_disk)
 
-        self._instance = testcloud.instance.Instance(
-            hostname=hostname,
-            image=self._image,
-            connection=f"qemu:///{self.connection}",
-            domain_configuration=self._domain)
+        with GuestTestcloud._testcloud_lock:
+            self._instance = testcloud.instance.Instance(
+                hostname=hostname,
+                image=self._image,
+                connection=f"qemu:///{self.connection}",
+                domain_configuration=self._domain)
         self.verbose('name', self.instance_name, 'green')
 
         # Decide if we want to multiply timeouts when emulating an architecture
@@ -721,19 +733,21 @@ class GuestTestcloud(tmt.GuestSsh):
         # Boot the virtual machine
         self.info('progress', 'booting...', 'cyan')
         assert libvirt is not None
-        try:
-            self._instance.prepare()
-            self._instance.spawn_vm()
-            self._instance.start(DEFAULT_BOOT_TIMEOUT * time_coeff)
-        except (testcloud.exceptions.TestcloudInstanceError,
-                libvirt.libvirtError) as error:
-            raise ProvisionError(
-                f'Failed to boot testcloud instance ({error}).')
-        self.guest = self._instance.get_ip()
-        self.port = int(self._instance.get_instance_port())
-        self.verbose('ip', self.guest, 'green')
-        self.verbose('port', self.port, 'green')
-        self._instance.create_ip_file(self.guest)
+
+        with GuestTestcloud._testcloud_lock:
+            try:
+                self._instance.prepare()
+                self._instance.spawn_vm()
+                self._instance.start(DEFAULT_BOOT_TIMEOUT * time_coeff)
+            except (testcloud.exceptions.TestcloudInstanceError,
+                    libvirt.libvirtError) as error:
+                raise ProvisionError(
+                    f'Failed to boot testcloud instance ({error}).')
+            self.guest = self._instance.get_ip()
+            self.port = int(self._instance.get_instance_port())
+            self.verbose('ip', self.guest, 'green')
+            self.verbose('port', self.port, 'green')
+            self._instance.create_ip_file(self.guest)
 
         # Wait a bit until the box is up
         if not self.reconnect(
@@ -756,22 +770,28 @@ class GuestTestcloud(tmt.GuestSsh):
         if self._instance and self.guest:
             self.debug(f"Stopping testcloud instance '{self.instance_name}'.")
             assert testcloud is not None
-            try:
-                self._instance.stop()
-            except testcloud.exceptions.TestcloudInstanceError as error:
-                raise tmt.utils.ProvisionError(
-                    f"Failed to stop testcloud instance: {error}")
+
+            with GuestTestcloud._testcloud_lock:
+                try:
+                    self._instance.stop()
+                except testcloud.exceptions.TestcloudInstanceError as error:
+                    raise tmt.utils.ProvisionError(
+                        f"Failed to stop testcloud instance: {error}")
+
             self.info('guest', 'stopped', 'green')
 
     def remove(self) -> None:
         """ Remove the guest (disk cleanup) """
         if self._instance:
             self.debug(f"Removing testcloud instance '{self.instance_name}'.")
-            try:
-                self._instance.remove(autostop=True)
-            except FileNotFoundError as error:
-                raise tmt.utils.ProvisionError(
-                    f"Failed to remove testcloud instance: {error}")
+
+            with GuestTestcloud._testcloud_lock:
+                try:
+                    self._instance.remove(autostop=True)
+                except FileNotFoundError as error:
+                    raise tmt.utils.ProvisionError(
+                        f"Failed to remove testcloud instance: {error}")
+
             self.info('guest', 'removed', 'green')
 
     def reboot(self,
@@ -786,7 +806,8 @@ class GuestTestcloud(tmt.GuestSsh):
             return super().reboot(hard=hard, command=command)
         if not self._instance:
             raise tmt.utils.ProvisionError("No instance initialized.")
-        self._instance.reboot(soft=not hard)
+        with GuestTestcloud._testcloud_lock:
+            self._instance.reboot(soft=not hard)
         return self.reconnect(timeout=timeout)

```

qcheng-redhat commented:
I tried the latest patch with libvirt-provision plugin, it works.  I got three guests provisioned parallelized. 

My plan:

execute:
    how: tmt

/plan:
    provision:
        - how: libvirt
          guestargs:
              --cpu host-model
              --ram=4096
              --vcpus=2
              --graphics vnc,listen=0.0.0.0
          distroname: xxxxxx
        - how: libvirt
          guestargs:
              --cpu host-model
              --ram=4096
              --vcpus=2
              --graphics vnc,listen=0.0.0.0
          distroname: xxxxxx
        - how: libvirt
          guestargs:
              --cpu host-model
              --ram=4096
              --vcpus=2
              --graphics vnc,listen=0.0.0.0
          distroname: xxxxxx
    discover:
        how: fmf

log:
04:47:17         queued provision task #1: default-0, default-1 and default-2
04:47:17
04:47:17         provision task #1: default-0, default-1 and default-2
04:47:17 [default-0]         started
04:47:17 [default-0]         how: libvirt
04:47:17 [default-1]         started
04:47:17 [default-0]         order: 50
04:47:17 [default-1]         how: libvirt
04:47:17 [default-2]         started
04:47:17 [default-1]         order: 50
04:47:17 [default-2]         how: libvirt
04:47:17 [default-2]         order: 50
...
04:54:53
04:54:53         summary: 3 guests provisioned

happz commented:
> I tried the latest patch with libvirt-provision plugin, it works. I got three guests provisioned parallelized.

Glad to hear that, thanks!

> /plan: provision: - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx discover: how: fmf

On another note: please, consider supporting `hardware` field, https://tmt.readthedocs.io/en/stable/spec/hardware.html, at least for the already supported requirements, `cpu` and `memory`. Plugins shipped with tmt would express part of the `guestargs` field from your example like this:

```yaml
hardware:
  cpu:
    cores: 2
  memory: 4096 MB
```

I know it's probably of no immediate benefit to you *now*, and I would fully understand the hesitation or focus on more needed features for your use cases, but in the long run, it would help end-users if plugins would speak the same language. The field is supported by tmt's own plugins, by Testing Farm, and we're slowly extending the requirements supported by individual plugins. I suppose your plugin is very similar to `virtual` in tmt repo, it would transform the requirements from `hardware` field into options, e.g. https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L489 demonstrates how we did this for `virtual` plugin. I'd be happy to provide any guidance if you decide t oextend your plugin.

qcheng-redhat commented:
> > I tried the latest patch with libvirt-provision plugin, it works. I got three guests provisioned parallelized.
> 
> Glad to hear that, thanks!
> 
> > /plan: provision: - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx - how: libvirt guestargs: --cpu host-model --ram=4096 --vcpus=2 --graphics vnc,listen=0.0.0.0 distroname: xxxxxx discover: how: fmf
> 
> On another note: please, consider supporting `hardware` field, https://tmt.readthedocs.io/en/stable/spec/hardware.html, at least for the already supported requirements, `cpu` and `memory`. Plugins shipped with tmt would express part of the `guestargs` field from your example like this:
> 
> ```yaml
> hardware:
>   cpu:
>     cores: 2
>   memory: 4096 MB
> ```
> 
> I know it's probably of no immediate benefit to you _now_, and I would fully understand the hesitation or focus on more needed features for your use cases, but in the long run, it would help end-users if plugins would speak the same language. The field is supported by tmt's own plugins, by Testing Farm, and we're slowly extending the requirements supported by individual plugins. I suppose your plugin is very similar to `virtual` in tmt repo, it would transform the requirements from `hardware` field into options, e.g. https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/testcloud.py#L489 demonstrates how we did this for `virtual` plugin. I'd be happy to provide any guidance if you decide t oextend your plugin.

Hi @happz , 

Thanks for the suggestion of using TMT hardware specification and sharing the `virtual` plugin example. Let's discuss this topic offline.

Thanks,
Qinghua

psss commented:
I've tried the parallel provision with the `/tests/multihost/web` test several times and it always failed with:

    Failed to connect in 60s.

Twice it was just a single box which failed and once even two of them. @happz, have you seen anything like this?

happz commented:
> I've tried the parallel provision with the `/tests/multihost/web` test several times and it always failed with:
> 
> ```
> Failed to connect in 60s.
> ```
> 
> Twice it was just a single box which failed and once even two of them. @happz, have you seen anything like this?

Nope, I haven't seen this error.

Well, I'm afraid we should push this one out into 1.31. The PR is changing the whole step, you're hitting issues I haven't seen, there's barely any review, and it's Nov 28th :/

happz commented:
/packit test --identifier full

happz commented:
/packit test --identifier full

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1819,6 +1824,120 @@ def show(self, keys: Optional[list[str]] = None) -> None:
+            # ... and then sit and wait as they get delivered to us as they
+            # finish. Unpack the guest and logger, so we could preserve logging
+            # and prepare the right outcome package.
+            for future in as_completed(futures):
```
There is no timeout on this level, it's up to individual plugins to give up. I'd say we're in a better situation than with other steps, actually, at least `provision` plugins do have some concept of timeouts, unlike the rest of the flock - I can imagine `prepare/shell` running an endless shell loop, and `prepare` knows nothing about deadlines or timeouts :)

happz commented on tmt/queue.py:
```diff
@@ -62,98 +45,86 @@ def name(self) -> str:
 
-        except Exception as exc:
-            # logger.info('finished', color='cyan')
+    # First, spawn all loggers, and set their labels if needed. Don't bother
```
Eye of the beholder, I guess... In this case, yes, it doesn't look weird, but in general, one sentence per line is not my favorite rule. Lucky for me, I think tmt does not have an official guidance like this :)

Fixed in 48da528d.

happz commented on tmt/steps/__init__.py:
```diff
@@ -1973,78 +1983,70 @@ def phase_name(self) -> str:
 
-@dataclasses.dataclass
-class PushTask(Task):
+@ dataclasses.dataclass
```
A typo. Fixed in 48da528d.

KwisatzHaderach replied:
there are also some spaces on property decorators, which you did not fix :)

happz replied:
Dammit. Fixed in ac122081

psss commented:
> > I've tried the parallel provision with the `/tests/multihost/web` test several times and it always failed with:
> > ```
> > Failed to connect in 60s.
> > ```
> > Twice it was just a single box which failed and once even two of them. @happz, have you seen anything like this?
> 
> Nope, I haven't seen this error.

Interesting. Tried today again and encountered similar behaviour. Seems that the problem happens randomly. Run four times:

* httpd-server failed to connect in 60s.
* httpd-server failed to connect in 60s.
* curl-client failed to connect in 60s.
* everything fine & tests passed

Looking through the log I see:

    err: Warning: Permanently added '192.168.124.110' (ED25519) to the list of known hosts.
    err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
    err: Permission denied, please try again.
    err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
    err: Permission denied, please try again.
    err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
    err: Received disconnect from 192.168.124.110 port 22:2: Too many authentication failures
    err: Disconnected from 192.168.124.110 port 22

Which is weird because the command line is the same as for the other guests:

    Run command: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oServerAliveInterval=60 -oServerAliveCountMax=5 -oIdentitiesOnly=yes -p22 -i /var/tmp/tmt/run-024/plan/provision/curl-client/id_ecdsa -S/run/user/12559/tmt/tmpyjjy9wtt root@192.168.124.110 'export TMT_PLAN_DATA=/var/tmp/tmt/run-024/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-024/plan/data/variables.env; export TMT_TREE=/var/tmp/tmt/run-024/plan/tree; export TMT_VERSION=1.30.0.dev64+g9df4fee9.d20240103; whoami'

Any idea what might be wrong?

happz commented:
> > > I've tried the parallel provision with the `/tests/multihost/web` test several times and it always failed with:
> > > ```
> > > Failed to connect in 60s.
> > > ```
> > > 
> > > Twice it was just a single box which failed and once even two of them. @happz, have you seen anything like this?
> > 
> > 
> > Nope, I haven't seen this error.
> 
> Interesting. Tried today again and encountered similar behaviour. Seems that the problem happens randomly. Run four times:
> 
> * httpd-server failed to connect in 60s.
> * httpd-server failed to connect in 60s.
> * curl-client failed to connect in 60s.
> * everything fine & tests passed
> 
> Looking through the log I see:
> 
> ```
> err: Warning: Permanently added '192.168.124.110' (ED25519) to the list of known hosts.
> err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
> err: Permission denied, please try again.
> err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
> err: Permission denied, please try again.
> err: ssh_askpass: exec(/usr/libexec/openssh/ssh-askpass): No such file or directory
> err: Received disconnect from 192.168.124.110 port 22:2: Too many authentication failures
> err: Disconnected from 192.168.124.110 port 22
> ```
> 
> Which is weird because the command line is the same as for the other guests:
> 
> ```
> Run command: ssh -oForwardX11=no -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oServerAliveInterval=60 -oServerAliveCountMax=5 -oIdentitiesOnly=yes -p22 -i /var/tmp/tmt/run-024/plan/provision/curl-client/id_ecdsa -S/run/user/12559/tmt/tmpyjjy9wtt root@192.168.124.110 'export TMT_PLAN_DATA=/var/tmp/tmt/run-024/plan/data; export TMT_PLAN_ENVIRONMENT_FILE=/var/tmp/tmt/run-024/plan/data/variables.env; export TMT_TREE=/var/tmp/tmt/run-024/plan/tree; export TMT_VERSION=1.30.0.dev64+g9df4fee9.d20240103; whoami'
> ```
> 
> Any idea what might be wrong?

Stinks like a race condition, threads contesting shared resources or something similar. I'll try to debug it in the afternoon.

psss commented:
> Stinks like a race condition, threads contesting shared resources or something similar. I'll try to debug it in the afternoon.

Thanks. The detailed debug output from `ssh` looks like this:

```
debug1: Offering public key: /home/psss/.ssh/id_rsa RSA SHA256:OY7RDp7mZHS6OzXvc9hnQk2qaDCcOuFM2iaGD8GOrKc explicit agent
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic,password
debug1: Offering public key: /home/psss/.ssh/1minutetip_rsa RSA SHA256:9j1blwt3wcrRiGYZQ7ZGu9axm3cDklH6/z4c+Ee8CzE explicit agent
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic,password
debug1: Offering public key: /var/tmp/tmt/run-031/plan/provision/wget-client/id_ecdsa ECDSA SHA256:i3ivQgMXnx8vh5PJ/L1K0LKqIs5RToHLZqA+d0QCbtM explicit
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic,password
debug2: we did not send a packet, disable method
debug3: authmethod_lookup password
debug3: remaining preferred: ,password
debug3: authmethod_is_enabled password
debug1: Next authentication method: password
...
```

Seems like the provided key is not accepted.

happz commented:
@psss hopefully addressed in 87ec95f5: added more locking in `testcloud` plugin - not all calls into `testcloud` library were protected, and added `_thread_safe` flag for provisioning plugins. Since we don't control all libraries used by our provisioning plugins, we get more time to verify they do work when used in parallel provisioning.

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+
+        def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[ActionTask]]:
+            queue: PhaseQueue[ProvisionStepData] = PhaseQueue(
+                'finish',
```
Should this be rather `action` or something like this?

happz replied:
Renamed in db0510dc, to `provision.action` and `provision.provision`.

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
 
-                    elif isinstance(phase, ProvisionPlugin):
-                        phase.go()
+        def _run_provision_phases(
```
Please, add a docstring, the return value is not very obvious.

happz replied:
Added in db0510dc

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+
+            return all_tasks, failed_tasks
+
+        def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[ActionTask]]:
```
Please, add a docstring, the return value is not very obvious.

happz replied:
Added in db0510dc

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+
+                    failed_tasks.append(outcome)
+
+                    continue
```
Not needed?

happz replied:
Fixed in db0510dc

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+            return all_tasks, failed_tasks
+
+        # Provisioning phases may be intermixed with actions. To perform all
+        # phases and action in a consistent manner, we will process them in
```
```suggestion
        # phases and actions in a consistent manner, we will process them in
```

happz replied:
Fixed in db0510dc

psss commented on tmt/steps/provision/__init__.py:
```diff
@@ -1918,45 +2045,147 @@ def go(self, force: bool = False) -> None:
+                                ProvisionPlugin[ProvisionStepData],
+                                all_phases.pop(0)))
+
+                _all_plugin_outcomes, _failed_plugin_outcomes = _run_provision_phases(
```
Why prefixing these variables with underscores?

happz replied:
Seems like a relict, this part used to be slightly different. Removed in db0510dc.

psss reviewed (changes requested):
Thanks much for implementing this! Looks good and works nicely. Added just some minor questions and suggestions.

psss commented:
This definitely deserves a mention in the release notes. Also, what do you think about adding a short paragraph with the overview of steps which support parallel execution to the Guide?

happz commented on tmt/queue.py:
```diff
@@ -13,24 +12,16 @@
-    #: A :py:class:`_Task` instace the outcome relates to.
-    task: TaskT
-
+class Task(Generic[TaskResultT]):
```
Added in db0510dc

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -47,6 +51,7 @@
     )
 
 if TYPE_CHECKING:
+
```
Removed in db0510dc

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1703,6 +1708,11 @@ class ProvisionPlugin(tmt.steps.GuestlessPlugin[ProvisionStepDataT]):
     _guest_class = Guest
 
+    #: If set, the plugin can be asked to provision in multiple threads at the
+    #: same time. Plugins that do no support parallel provisioning should keep
```
Fixed in db0510dc

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1819,6 +1829,120 @@ def show(self, keys: Optional[list[str]] = None) -> None:
 
+@dataclasses.dataclass
+class ProvisionTask(tmt.queue.GuestlessTask[None]):
+    """ A task to run provisioning of multiple guest """
```
Fixed in db0510dc

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1819,6 +1829,120 @@ def show(self, keys: Optional[list[str]] = None) -> None:
+    phases: list[ProvisionPlugin[ProvisionStepData]]
+
+    #: When ``ProvisionTask`` instance is received from the queue, ``phase``
+    #: points to the phase that has been provisioned for by the task.
```
Fixed in db0510dc

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -1819,6 +1829,120 @@ def show(self, keys: Optional[list[str]] = None) -> None:
+                if multiple_guests:
+                    new_logger.info('started', color='cyan')
+
+                # Submit each task/guest combination (save the guest & logger
```
Fixed in db0510dc

happz commented:
> This definitely deserves a mention in the release notes.

Sure, added one in a8d369a2.

> Also, what do you think about adding a short paragraph with the overview of steps which support parallel execution to the Guide?

I mentioned compatible plugins in the release note, and I was hoping to make this information available on plugin pages, once https://github.com/teemtee/tmt/pull/2549 lands

psss commented:
> > Also, what do you think about adding a short paragraph with the overview of steps which support parallel execution to the Guide?
> 
> I mentioned compatible plugins in the release note, and I was hoping to make this information available on plugin pages, once #2549 lands

Makes sense, thanks.

psss reviewed (approved):
Thanks for addressing all issues. Looks good now. I've just slightly modified the release note formatting in 59681ab2.

psss commented:
Many tests are failing because of:

    PullTask.__init__() got an unexpected keyword argument 'result'

happz commented:
/packit test --identifier full

> Many tests are failing because of:
> 
> ```
> PullTask.__init__() got an unexpected keyword argument 'result'
> ```

Adding b6a3250a, let's see whether it helps.

psss commented:
/packit test --identifier full

psss commented:
/packit test --identifier full

guoguojenna commented:
Seems beaker provisioning is still in sequence, any reasons?

Thanks.

guoguojenna commented on tmt/steps/provision/mrack.py:
```diff
@@ -680,6 +680,8 @@ class ProvisionBeaker(tmt.steps.provision.ProvisionPlugin[ProvisionBeakerData]):
     _data_class = ProvisionBeakerData
     _guest_class = GuestBeaker
 
+    # _thread_safe = True
```
Uncomment it, use _thread_safe = True, beaker parallel provisioning works for me. Wondering why you comment it?

[client-1]         guest: has been requested
[client-1]         job id: 8784929
[client-1]         wait: waiting for condition 'get_new_state' with timeout 1:00:00, deadline in 3600.0 seconds, checking every 60.00 seconds
[server-1]         guest: has been requested
[server-1]         job id: 8784930
[server-1]         wait: waiting for condition 'get_new_state' with timeout 1:00:00, deadline in 3600.0 seconds, checking every 60.00 seconds
[client-2]         guest: has been requested
[client-2]         job id: 8784931
...
summary: 3 guests provisioned
            client-2
            client-1
            server-1

happz replied:
Because we are not sure whether the library tmt uses for Beaker provisioning, mrack, is thread-safe. We did run into issues with the `virtual` plugin, where the problem was eventually solved by adding an extra lock to provide serialization, but there was no time invested into doing the same for the Beaker plugin. If it's required, that is.

Filed https://github.com/teemtee/tmt/issues/2607 to track this.
